### PR TITLE
🐛Fixed image properties to be reset to null after removal

### LINKED ISSUE #9085: Deleting a feature image becomes "" instead of null and breaks filter query
Hello,
I was trying to get all posts that have set a featured image.

`{{#get "posts" limit="5" filter="id:-{{id}}+feature_image:-null"}}`

First ... I imported all the content from another blog. And the rule `feature_image:-null` was ignored.
I created a new post without the featured image and the filter works. It doesn't get the post. Then I set a featured image, saved the post, removed the featured image and saved again the post. Now the rule `feature_image:-null` gets ignored and the query selects the new post, even if it doesn't have an image.

So something is set in db instead of `null` after the featured image gets removed. 

I exported the json and checked the `feature_image` value.
When I created the post the value was `"feature_image": null` and it was good, but after I set the image and removed it, it became `"feature_image": ""`.

This breaks the filter query.

P.S. In this url: https://api.ghost.org/docs/filter ... it says `filter="featured:true,image:-null"` and should, probably, be `filter="featured:true,feature_image:-null"` and `images:-null` shoud be `feature_image:-null`

* Ghost Version: 1.10.0

### PR DESCRIPTION
**Pull request by aileen**
closes #9085

Fixes an issue, where the client sets image properties to `""` after deleting the image. This causes problems with the query filter (see TryGhost/GQL#24), as they have to be `null`.

Added a check in the model layer saving method to set value  to `null`, when the property is empty.

Affected models and properties:
- `posts`:
	- `feature_image`
	- `og_image`
	- `twitter_image`
- `users`:
	- `profile_image`
	- `cover_image`
- `tags`:
	- `feature_image`

Note: This is very rare edge case and we therefore decided to not write a migration for it. @kirrg001 Please correct me, if I'm wrong.

### CODE REVIEW

**Comment by AileenCGN** [comment]
@kirrg001 wait... more is coming. Just saw this line [here](https://github.com/TryGhost/Ghost/issues/9085#issuecomment-333819526) now:

> This affects all images, because the admin client sends an empty string instead of null

**Comment by AileenCGN** [comment]
@kirrg001 Ready for review now!

**Comment by kirrg001** [comment]
Could you please use unit tests? 😁

**Comment by AileenCGN** [comment]
😱 I will try...

File: core/server/models/post.js [hunk 1]
```diff
@@ -163,6 +163,21 @@ Post = ghostBookshelf.Model.extend({
             tagsToSave,
             ops = [];
 
+        // CASE: the client might send empty image properties with "" instead of setting them to null.
+        // This can cause GQL to fail. We therefore enforce 'null' for empty image properties.
+        // See https://github.com/TryGhost/GQL/issues/24
+        if (this.get('feature_image') === '' && this.previous('feature_image')) {
```
Comment (kirrg001) [inline core/server/models/post.js]: Can't we use `onValidate`?

`onValidate` is always called after `onSaving`.

Post
```
onValidate
   validate '' fields
   call parent
```

Comment (AileenCGN) [reply core/server/models/post.js]: Alright. Done that. Also used a generic fn now. Will change the tests now.

File: core/test/utils/fixtures/data-generator.js [hunk 2]
```diff
@@ -32,7 +32,8 @@ DataGenerator.Content = {
...
             published_at: new Date("2015-01-01"),
-            custom_excerpt: 'This is my custom excerpt!'
+            custom_excerpt: "This is my custom excerpt!",
```
Comment (AileenCGN) [inline core/test/utils/fixtures/data-generator.js]: @kirrg001 is there a reason why we use mixed quotes in data-generator? Sometimes we use single and sometimes double quotes. I was a bit confused about that...

File: core/test/unit/models/base/base_spec.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    afterEach(function () {
+        sandbox.restore();
+    });
+
+    describe('setEmptyValuesToNull', function () {
+        it('should add function to prototype', function () {
+            ghostBookshelf.Model.prototype.setEmptyValuesToNull.should.be.a.Function();
+        });
+        it.skip('resets given empty value to null', function (done) {
```
Comment (AileenCGN) [inline core/test/unit/models/base/base_spec.js]: @kirrg001 This is giving me a headache... Not sure how I could test the functionality here... Can you help?

File: core/test/utils/fixtures/data-generator.js (hunk 2 above)
Comment (kirrg001) [reply core/test/utils/fixtures/data-generator.js]: No reason, just a mistake 💇🏼‍♂️

File: core/test/unit/models/base/base_spec.js (hunk 3 above)
Comment (kirrg001) [reply core/test/unit/models/base/base_spec.js]: What would you like to achieve and what isn't working?

Comment (AileenCGN) [reply core/test/unit/models/base/base_spec.js]: I want to unit test the fn itself. Do you think it's necessary?

File: core/test/unit/models/post_spec.js [hunk 4]
```diff
@@ -2,15 +2,48 @@ var should = require('should'), // jshint ignore:line
...
     before(function () {
         models.init();
     });
 
+    after(function () {
+        sandbox.restore();
+    });
+
+    describe('onValidate', function () {
+        let knexMock;
```
Comment (AileenCGN) [inline core/test/unit/models/post_spec.js]: @kirrg001 Travis is failing here for Node v4. Do you know why it worked for you?

File: core/test/unit/models/base/base_spec.js (hunk 3 above)
Comment (kirrg001) [reply core/test/unit/models/base/base_spec.js]: > and what isn't working?

Comment (AileenCGN) [reply core/test/unit/models/base/base_spec.js]: I wasn't able to stub the `get()` fn for the bookshelf model as it is needed [here](https://github.com/TryGhost/Ghost/pull/9432/files#diff-f8d02ad12b13d44b912bbf46cf7193b6R292)...

Comment (kirrg001) [reply core/test/unit/models/base/base_spec.js]: You have to bind the model instance.
Right now you are binding the test context.

File: core/test/unit/models/post_spec.js (hunk 4 above)
Comment (AileenCGN) [reply core/test/unit/models/post_spec.js]: See https://github.com/TryGhost/Ghost/blob/master/core/test/unit/models/user_spec.js#L13

Comment (AileenCGN) [reply core/test/unit/models/post_spec.js]: nvm, found the issue 😂

File: core/server/models/post.js [hunk 5]
```diff
@@ -302,6 +302,17 @@ Post = ghostBookshelf.Model.extend({
         return sequence(ops);
     },
 
+    onValidate: function onValidate() {
+        // CASE: the client might send empty image properties with "" instead of setting them to null.
+        // This can cause GQL to fail. We therefore enforce 'null' for empty image properties.
+        // See https://github.com/TryGhost/GQL/issues/24
+        var affectedProps = ['feature_image', 'og_image', 'twitter_image'];
+
+        ghostBookshelf.Model.prototype.setEmptyValuesToNull.call(this, affectedProps);
```
Comment (kirrg001) [inline core/server/models/post.js]: I would suggest to move the logic into the parent/base fn. We currently have to override the base fn three times and it's doing the same.

```
onValidate: function onValidate(model, columns, options) {
    this.careAboutEmptyString(...);
    return validation.validateSchema(this.tableName, this, options);
}

careAboutEmptyString: fn
  if (!this.emptyStringFields)
     return;

  const properties = this.emptyStringFields();
  ...
```

e.g. in Post model

```
emptyStringFields: fn
   return ['...'];
```

Naming of functions is just an example and not good enough.

Comment (kirrg001) [reply core/server/models/post.js]: This has also the advantage, that you can easily add the functionality for other models in the future (without overriding the validate fn)

Comment (AileenCGN) [reply core/server/models/post.js]: Makes sense to me!

File: core/test/unit/models/base/base_spec.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        before(function () {
+            knexMock = new testUtils.mocks.knex();
+            knexMock.mock();
+        });
+
+        after(function () {
+            knexMock.unmock();
+        });
+
+        // TODO: mock the base model und run those tests
```
Comment (AileenCGN) [inline core/test/unit/models/base/base_spec.js]: @kirrg001 I am done with the rest, except for this. I couldn't figure out how to test the base model functionality... But the functionality itself is tested within the post, user, and tags models.

Comment (kirrg001) [reply core/test/unit/models/base/base_spec.js]: pushed a commit 👍

**Pull request merged**